[careers] Export structured data for Google Jobs.

nisarhassan12 opened the pull request:
Fixes gitpod-io/website#696

nisarhassan12 commented:
seems to be working https://search.google.com/test/rich-results?utm_campaign=sdtt&utm_medium=url&id=jh-BpIcq5h-GQjl2Ui-srA

![image](https://user-images.githubusercontent.com/46004116/88013927-1558c380-cb37-11ea-8765-5ef2a5e35259.png)

Not sure what the `datePosted` is gonna be.

jankeromnes commented on src/components/careers/ExpandableJob.tsx:
```diff
@@ -76,6 +77,52 @@ const ExpandableJob = ({ title, intro, paragraphs, lists, textAfterTheLists, ren
+          "@context": "https://schema.org/",
+          "@type": "JobPosting",
+          "title": title,
+          "description": `<p>${intro}</p>${paragraphs.split('\n').map((p, i) => <p key={i+title}>{p}</p>)}`,
```
There seems to be a problem with the descriptions for both job offers:

The Developer Tools Engineer description is:

> `<p>If you were ever frustrated with a development tool you used and decided to build a better one yourself, this offer is for you.</p>[object Object],[object Object]`

Most of the description is missing, and the `[object Object]` seem to be bugs.

The UX Developer description is:

> `<p>Gitpod provides ready-to-code development environments for professional software engineering teams. We set out to change how developers work by eliminating tedious and repetitive configuration of local machines and replacing it with a fully integrated, cloud-based solution. Gitpod is the missing piece in a modern DevOps toolchain.</p>[object Object]`

This doesn't seem to be relevant for UX developers, and there is also the `[object Object]` bug.

nisarhassan12 commented:
@jankeromnes I have fixed the `descriptions` and added the dates. Please have a look whenever you'll have time. Thanks 
![image](https://user-images.githubusercontent.com/46004116/92855561-d98dfc00-f40b-11ea-98f0-08ec22e17605.png)

jankeromnes commented:
Many thanks @nisarhassan12! I'll take another look.

Also, since this is ready for review, please remove the `wip: ` prefix from the Pull Request title. 🙏

nisarhassan12 commented:
> Also, since this is ready for review, please remove the wip: prefix from the Pull Request title. pray

@jankeromnes Thanks. Done :)

jankeromnes commented on src/components/careers/ExpandableJob.tsx:
```diff
@@ -3,112 +3,166 @@ import React, { useState } from 'react'
+    '@context': 'https://schema.org/',
+    '@type': 'JobPosting',
+    title: title,
+    description: `<p>${intro}</p>${paragraphs.split('\n').map((p) => `<p>${p}</p>`)}`,
```
I think you also want to include the `lists` and the `textAfterTheLists` in this `description`. (Note: You may use `<ul>` and `<li>` for the lists.)

FYI, according to https://developers.google.com/search/docs/data-types/job-posting, here are the guidelines for what a good `description` should be:

> The full description of the job in HTML format.
> 
> The `description` should be a complete representation of the job, including job responsibilities, qualifications, skills, working hours, education requirements, and experience requirements. The `description` can't be the same as the `title`.
> 
> Additional guidelines:
> 
> - You must format the description in HTML.
> - At minimum, add paragraph breaks using `<br>`, `<p>`, or `\n`.
> - The feature recognizes the following HTML tags: `<p>`, `<ul>`, and `<li>`.
> - The feature doesn't recognize header and character-level tags, such as `<h1>`, `<strong>`, and `<em>`. While the tags won't affect the formatting in the feature, you can safely include them on the page.

jankeromnes commented on src/components/careers/ExpandableJob.tsx:
```diff
@@ -3,112 +3,166 @@ import React, { useState } from 'react'
+    title: title,
+    description: `<p>${intro}</p>${paragraphs.split('\n').map((p) => `<p>${p}</p>`)}`,
+    datePosted: date,
+    hiringOrganization: {
+      '@type': 'Organization',
+      name: 'TypeFox GmbH',
+      sameAs: 'https://www.typefox.io/',
+      logo: TypeFoxLogo,
+      contactPoint: {
+        '@type': 'ContactPoint',
+        url: 'https://www.typefox.io/contact/'
+      }
+    },
```
I think, going forward, we'll want Gitpod's hiring organization to be "Gitpod Inc." instead of "TypeFox GmbH":

```suggestion
    hiringOrganization: {
      '@type': 'Organization',
      name: 'Gitpod Inc.',
      sameAs: 'https://www.gitpod.io/',
      logo: GitpodLogo,
      contactPoint: {
        '@type': 'ContactPoint',
        url: 'https://www.gitpod.io/contact/'
      }
    },
```

jankeromnes commented on src/components/careers/ExpandableJob.tsx:
```diff
@@ -3,112 +3,166 @@ import React, { useState } from 'react'
+      '@type': 'Organization',
+      name: 'TypeFox GmbH',
+      sameAs: 'https://www.typefox.io/',
+      logo: TypeFoxLogo,
```
Also, please check that this logo isn't being used anywhere else, because it doesn't look good:

<img width="278" alt="Screenshot 2020-09-11 at 11 07 28" src="https://user-images.githubusercontent.com/599268/92901095-82deed00-f420-11ea-896a-7bcffe1ba37c.png">

- It's blurry & pixellated (because it's PNG and not SVG)
- Maybe it should have the Fox as well?
- In any case, for the hiring organization, we should use a Gitpod logo instead (please see other comment)

nisarhassan12 replied:
We are using it on the about page and the Fox and the word TypeFox are two different images.
![image](https://user-images.githubusercontent.com/46004116/95648462-d6605b80-0af0-11eb-9673-36deb02fb842.png)

nisarhassan12 replied:
@ChristinFrohne Can you provide me with both logos as a single image? Thanks

jankeromnes reviewed (commented):
Many thanks for the great work here!

Just a few "minor" adjustments and this will be good to go! 👍 I personally can't wait for the job offer to make it to Google Jobs, because we really need a talented / experienced UX Developer. I really hope Google can help us find one. 🙂

nisarhassan12 commented:
@jankeromnes I have made the requested changes. Please have a look whenever you'll have time. Thanks

ChristinFrohne commented:
Is this still relevant? Does it need to be adjusted for every new job opening or would it work automatically?

nisarhassan12 commented:
Thanks, @ChristinFrohne Yes this is still relevant and It would work automatically for every new job posting.

jankeromnes commented:
Many thanks for making this work @nisarhassan12! (And sorry for the long review times!)

As discussed in our 1:1, the results now look quite good:

<img width="789" alt="Screenshot 2021-01-19 at 11 00 11" src="https://user-images.githubusercontent.com/599268/105018741-a573b900-5a45-11eb-95fa-00207eca81bc.png">

Also, when you try to render the description HTML markup (e.g. by copying the HTML, then opening a new browser tab, typing `data:text/html,` and then pasting the entire description HTML markup), it looks quite good:

<img width="1440" alt="Screenshot 2021-01-19 at 10 14 04" src="https://user-images.githubusercontent.com/599268/105018900-cf2ce000-5a45-11eb-9fa4-2a18655c6f86.png">

The only problem is that there is a single isolated comma (`,`) on some lines. You can even see it in the HTML markup:

<img width="599" alt="Screenshot 2021-01-19 at 10 15 43" src="https://user-images.githubusercontent.com/599268/105018970-e10e8300-5a45-11eb-8339-ef4d6a427ad7.png">

I'm not yet sure why that is, but I suspect that maybe in some places you're trying to render an array, and for some reason JSX stringifies your array by adding a `,` between each element. In that case, you could try finding the arrays you're rendering, and then adding a single `.join(' ')` after them (e.g. like so `text.split('\n').map(l => '<p>' + l + '</p>').join(' ')`).

nisarhassan12 commented:
Thanks @jankeromnes for the review I was able to fix the issue via using .join(' ') at the end of all of the maps.
![image](https://user-images.githubusercontent.com/46004116/105050632-cfea6400-5a8f-11eb-88a8-d7a9a85fcc2b.png)

nisarhassan12 commented:
Merging this as it seems to work well.
